### PR TITLE
(PUP-3899) Testing for checksum type on file resource

### PR DESCRIPTION
**Pull request by MikaelSmith**
Add integration and acceptance tests for specifying the checksum on a file. Covers puppet apply, querying the server via the HTTP API, and puppet agent runs.

### CODE REVIEW

**Comment by MikaelSmith** [comment]
TODO:
* Depends on ctime/mtime comparison fix for source_attribute to pass.
* Squash last commit.
* Update integration test to not re-use the catalog.

**Comment by MikaelSmith** [comment]
Oops, ended up with 2 other commits.

File: spec/integration/network/http/api/indirected_routes_spec.rb [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+
+        it "should find the file metadata with expected checksum" do
+          request = a_request_that_finds(Puppet::IndirectorProxy.new("modules/lib/file.rb", "file_metadata"),
+                                         {:accept_header => 'unknown, text/pson'},
+                                         {:environment => 'production', :checksum_type => checksum_type})
+          handler.call(request, response)
+          resp = JSON.parse(response.body)
+
+          expect(resp['checksum']['type']).to eq(checksum_type)
+          expect(resp['checksum']['value']).to eq("{#{checksum_type}}#{checksum}")
```
Comment (joshcooper) [inline spec/integration/network/http/api/indirected_routes_spec.rb]: This looks to be failing due to `eq` with ctime/mtime:

```
  1) Puppet::Network::HTTP::API::IndirectedRoutes when running the master application using Puppet API to request file metadata when checksum_type is ctime should find the file metadata with expected checksum
     Failure/Error: expect(resp['checksum']['value']).to eq("{#{checksum_type}}#{checksum}")

       expected: "{ctime}2015-02-05 15:11:56 -0800"
            got: "{ctime}2015-02-05 15:12:00 -0800"

       (compared using ==)
     # ./spec/integration/network/http/api/indirected_routes_spec.rb:37:in `block (5 levels) in <top (required)>'
     # util/rspec_runner:39:in `run'
     # util/rspec_runner:53:in `<main>'
  2) Puppet::Network::HTTP::API::IndirectedRoutes when running the master application using Puppet API to request file metadata when checksum_type is ctime should search for the file metadata with expected checksum
     Failure/Error: expect(file['checksum']['value']).to eq("{#{checksum_type}}#{checksum}")

       expected: "{ctime}2015-02-05 15:11:56 -0800"
            got: "{ctime}2015-02-05 15:12:00 -0800"

       (compared using ==)
     # ./spec/integration/network/http/api/indirected_routes_spec.rb:51:in `block (5 levels) in <top (required)>'
     # util/rspec_runner:39:in `run'
     # util/rspec_runner:53:in `<main>'
  3) Puppet::Network::HTTP::API::IndirectedRoutes when running the master application using Puppet API to request file metadata when checksum_type is mtime should find the file metadata with expected checksum
     Failure/Error: expect(resp['checksum']['value']).to eq("{#{checksum_type}}#{checksum}")

       expected: "{mtime}2015-02-05 15:11:56 -0800"
            got: "{mtime}2015-02-05 15:12:00 -0800"

       (compared using ==)
     # ./spec/integration/network/http/api/indirected_routes_spec.rb:37:in `block (5 levels) in <top (required)>'
     # util/rspec_runner:39:in `run'
     # util/rspec_runner:53:in `<main>'
  4) Puppet::Network::HTTP::API::IndirectedRoutes when running the master application using Puppet API to request file metadata when checksum_type is mtime should search for the file metadata with expected checksum
     Failure/Error: expect(file['checksum']['value']).to eq("{#{checksum_type}}#{checksum}")

       expected: "{mtime}2015-02-05 15:11:56 -0800"
            got: "{mtime}2015-02-05 15:12:00 -0800"

       (compared using ==)
     # ./spec/integration/network/http/api/indirected_routes_spec.rb:51:in `block (5 levels) in <top (required)>'
     # util/rspec_runner:39:in `run'
     # util/rspec_runner:53:in `<main>'
Finished in 16.08 seconds (files took 3.42 seconds to load)
```

File: spec/integration/network/http/api/indirected_routes_spec.rb [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+                                            {:accept_header => 'unknown, text/pson'},
+                                            {:environment => 'production', :checksum_type => checksum_type, :recurse => 'yes'})
+          handler.call(request, response)
+          resp = JSON.parse(response.body)
+
+          expect(resp.length).to eq(2)
+          file = resp.find {|x| x['relative_path'] == 'file.rb'}
+
+          expect(file['checksum']['type']).to eq(checksum_type)
+          expect(file['checksum']['value']).to eq("{#{checksum_type}}#{checksum}")
```
Comment (joshcooper) [inline spec/integration/network/http/api/indirected_routes_spec.rb]: same here

**Comment by MikaelSmith** [comment]
Updated to share the checksum validity check from shared_contexts/checksum.rb.

**Pull request merged**